# [Pandas](https://pandas.pydata.org/): Python Data Analysis Library

##### (doesn't quite match up, but it's still a cute nickname!)

<sub>(According to the Wikipedia page on the Pandas library, the name is derived from panel data, but I think we get the point by now)</sub>

## What can it do?

Pandas can:
1. Import data from raw text files (like .csv files), or from Excel files (.xls or .xlsx) into Python
2. Perform data manipulation tasks on this imported data, such as
    1. changing column names, 
    2. selecting specific columns as per our needs, 
    3. slicing the rows into multiple parts, 
    4. changing indices, 
    5. joining or merging data, 
    6. and much more.
3. Export this manipulated data to numpy arrays for numerical and statistical analysis.

In [2]:
import pandas as pd

In [3]:
type(pd)

module

In [4]:
print(pd)

<module 'pandas' from '/home/abhigyan/anaconda3/lib/python3.7/site-packages/pandas/__init__.py'>


# Importing Data - The DataFrame:

##### Trial 1

In [16]:
#ip = pd.read_csv("")
data = pd.read_csv("Zoo.csv",delimiter=',')

Now, what is this `data`? What data type is it and what kind of information does it store? Also, how do we see the data stored in it?

In [6]:
print(type(data))

<class 'pandas.core.frame.DataFrame'>


Thus, the basic data type of the Pandas library is the `DataFrame`, which is essentially a table of values, similar to those you get in Excel or other spreadsheet applications.

# Displaying Data
To display the first 5 rows of data, we use `head()`

In [7]:
data.head(50)

animal  uniq_id  water_need
0   elephant     1001         500
1   elephant     1002         600
2   elephant     1003         550
3      tiger     1004         300
4      tiger     1005         320
5      tiger     1006         330
6      tiger     1007         290
7      tiger     1008         310
8      zebra     1009         200
9      zebra     1010         220
10     zebra     1011         240
11     zebra     1012         230
12     zebra     1013         220
13     zebra     1014         100
14     zebra     1015          80
15      lion     1016         420
16      lion     1017         600
17      lion     1018         500
18      lion     1019         390
19  kangaroo     1020         410
20  kangaroo     1021         430
21  kangaroo     1022         410

To display the last 5 rows of data, we use `tail()`

In [8]:
data.tail()

animal  uniq_id  water_need
17      lion     1018         500
18      lion     1019         390
19  kangaroo     1020         410
20  kangaroo     1021         430
21  kangaroo     1022         410

To display `n` random lines, we use the `sample(n)` function:

In [9]:
data.sample(10)

animal  uniq_id  water_need
17      lion     1018         500
4      tiger     1005         320
0   elephant     1001         500
18      lion     1019         390
11     zebra     1012         230
19  kangaroo     1020         410
6      tiger     1007         290
12     zebra     1013         220
15      lion     1016         420
7      tiger     1008         310

To display the entire data, we just write the name of our DataFrame object. (Note that it will only show the entire table if it is small enough to be displayed, and we can set that using the `display.max_rows` variable)

In [10]:
data

animal  uniq_id  water_need
0   elephant     1001         500
1   elephant     1002         600
2   elephant     1003         550
3      tiger     1004         300
4      tiger     1005         320
5      tiger     1006         330
6      tiger     1007         290
7      tiger     1008         310
8      zebra     1009         200
9      zebra     1010         220
10     zebra     1011         240
11     zebra     1012         230
12     zebra     1013         220
13     zebra     1014         100
14     zebra     1015          80
15      lion     1016         420
16      lion     1017         600
17      lion     1018         500
18      lion     1019         390
19  kangaroo     1020         410
20  kangaroo     1021         430
21  kangaroo     1022         410

In [11]:
pd.get_option("display.max_rows")

60

Since our table has only 21 rows, it will show the entire thing at once (since `max_rows = 60`), but as we will see below, when it becomes very large, then the entire table won't be displayed.

Now, we will use another dataset which is considerably larger, and uses a different delimiter:

# Importing more Data

##### Trial 2

In [14]:
data2 = pd.read_csv("pandas_tutorial_read.csv",delimiter=";")

In [15]:
data2

2018-01-01 00:01:01  read  country_7  2458151261      SEO  North America
0     2018-01-01 00:03:20  read  country_7  2458151262      SEO  South America
1     2018-01-01 00:04:01  read  country_7  2458151263  AdWords         Africa
2     2018-01-01 00:04:02  read  country_7  2458151264  AdWords         Europe
3     2018-01-01 00:05:03  read  country_8  2458151265   Reddit  North America
4     2018-01-01 00:05:42  read  country_6  2458151266   Reddit  North America
...                   ...   ...        ...         ...      ...            ...
1789  2018-01-01 23:57:14  read  country_2  2458153051  AdWords  North America
1790  2018-01-01 23:58:33  read  country_8  2458153052      SEO           Asia
1791  2018-01-01 23:59:36  read  country_6  2458153053   Reddit           Asia
1792  2018-01-01 23:59:36  read  country_7  2458153054  AdWords         Europe
1793  2018-01-01 23:59:38  read  country_5  2458153055   Reddit           Asia

[1794 rows x 6 columns]

However, this one doesn't even contain headers... That's an issue, so we will now add those ourselves:

In [17]:
data2 = pd.read_csv("pandas_tutorial_read.csv", delimiter=";", names=['datetime','event','country','user_id','source','topic'])

In [21]:
print(data2)
data2

                 datetime event    country     user_id   source          topic
0     2018-01-01 00:01:01  read  country_7  2458151261      SEO  North America
1     2018-01-01 00:03:20  read  country_7  2458151262      SEO  South America
2     2018-01-01 00:04:01  read  country_7  2458151263  AdWords         Africa
3     2018-01-01 00:04:02  read  country_7  2458151264  AdWords         Europe
4     2018-01-01 00:05:03  read  country_8  2458151265   Reddit  North America
...                   ...   ...        ...         ...      ...            ...
1790  2018-01-01 23:57:14  read  country_2  2458153051  AdWords  North America
1791  2018-01-01 23:58:33  read  country_8  2458153052      SEO           Asia
1792  2018-01-01 23:59:36  read  country_6  2458153053   Reddit           Asia
1793  2018-01-01 23:59:36  read  country_7  2458153054  AdWords         Europe
1794  2018-01-01 23:59:38  read  country_5  2458153055   Reddit           Asia

[1795 rows x 6 columns]


datetime event    country     user_id   source          topic
0     2018-01-01 00:01:01  read  country_7  2458151261      SEO  North America
1     2018-01-01 00:03:20  read  country_7  2458151262      SEO  South America
2     2018-01-01 00:04:01  read  country_7  2458151263  AdWords         Africa
3     2018-01-01 00:04:02  read  country_7  2458151264  AdWords         Europe
4     2018-01-01 00:05:03  read  country_8  2458151265   Reddit  North America
...                   ...   ...        ...         ...      ...            ...
1790  2018-01-01 23:57:14  read  country_2  2458153051  AdWords  North America
1791  2018-01-01 23:58:33  read  country_8  2458153052      SEO           Asia
1792  2018-01-01 23:59:36  read  country_6  2458153053   Reddit           Asia
1793  2018-01-01 23:59:36  read  country_7  2458153054  AdWords         Europe
1794  2018-01-01 23:59:38  read  country_5  2458153055   Reddit           Asia

[1795 rows x 6 columns]

That looks right...

(Note 2: If you are wondering what’s in this data set – this is the data log of a travel blog taken from [data36.com](https://data36.com/pandas-tutorial-1-basics-reading-data-files-dataframes-data-selection/).)

# Importing Excel Spreadsheets:

This might be a super handy tool for those of you that use Excel spreadsheets a lot, so lets see this once as well:

(I've just saved the same values as those that we had in our `zoo.csv` file, so it's exactly the same thing, but this shows the principle of importing Excel into Python, and now you can try this with your own data as well)

In [29]:
data3 = pd.read_excel("Zoo.xlsx")
data3

checkdata = (data3 == data)

print(type(checkdata))

<class 'pandas.core.frame.DataFrame'>


# Choosing certain specific columns of your data:

In [30]:
data2[['user_id','country']]

user_id    country
0     2458151261  country_7
1     2458151262  country_7
2     2458151263  country_7
3     2458151264  country_7
4     2458151265  country_8
...          ...        ...
1790  2458153051  country_2
1791  2458153052  country_8
1792  2458153053  country_6
1793  2458153054  country_7
1794  2458153055  country_5

[1795 rows x 2 columns]

In [32]:
data2[['user_id','country']]

user_id    country
0     2458151261  country_7
1     2458151262  country_7
2     2458151263  country_7
3     2458151264  country_7
4     2458151265  country_8
...          ...        ...
1790  2458153051  country_2
1791  2458153052  country_8
1792  2458153053  country_6
1793  2458153054  country_7
1794  2458153055  country_5

[1795 rows x 2 columns]

Note that we need to put this within a double set of brackets. This is because the outer brackets tell python that you wish to select columns, and the inner one is for the actual list of columns that you wish to choose.

So, it is like this:

DataFrame\[list\], where list = \['column1','column2'\].

# Series
Now, we will see another Pandas data type, the `Series`. It is nothing but a one-column table, and it is useful in some cases which we will also come across soon.

If we want a series object instead of a `DataFrame`, we can do it in 2 ways:

`DataFrame.columnName` or `DataFrame['columnName']`

Examples follow below:

In [33]:
data.animal

0     elephant
1     elephant
2     elephant
3        tiger
4        tiger
5        tiger
6        tiger
7        tiger
8        zebra
9        zebra
10       zebra
11       zebra
12       zebra
13       zebra
14       zebra
15        lion
16        lion
17        lion
18        lion
19    kangaroo
20    kangaroo
21    kangaroo
Name: animal, dtype: object

In [34]:
data['animal']

0     elephant
1     elephant
2     elephant
3        tiger
4        tiger
5        tiger
6        tiger
7        tiger
8        zebra
9        zebra
10       zebra
11       zebra
12       zebra
13       zebra
14       zebra
15        lion
16        lion
17        lion
18        lion
19    kangaroo
20    kangaroo
21    kangaroo
Name: animal, dtype: object

# Filtering out specific rows based on values or other info:

Let’s say, you want to see a list of only the users who came from the ‘SEO’ source. In this case you have to filter for the ‘SEO’ value in the ‘source’ column:

In [35]:
data2[data2.source=='SEO']

datetime event    country     user_id source          topic
0     2018-01-01 00:01:01  read  country_7  2458151261    SEO  North America
1     2018-01-01 00:03:20  read  country_7  2458151262    SEO  South America
11    2018-01-01 00:08:57  read  country_7  2458151272    SEO      Australia
15    2018-01-01 00:11:22  read  country_7  2458151276    SEO  North America
16    2018-01-01 00:13:05  read  country_8  2458151277    SEO  North America
...                   ...   ...        ...         ...    ...            ...
1772  2018-01-01 23:45:58  read  country_7  2458153033    SEO  South America
1777  2018-01-01 23:49:52  read  country_5  2458153038    SEO  North America
1779  2018-01-01 23:51:25  read  country_4  2458153040    SEO  South America
1784  2018-01-01 23:54:03  read  country_2  2458153045    SEO  North America
1791  2018-01-01 23:58:33  read  country_8  2458153052    SEO           Asia

[346 rows x 6 columns]

So, how did this work? Let us understand clearly:

What we put inside the brackets is basically going to return a `Series` of boolean values telling us if the `source` column of that row contains `'SEO'` or not:

In [36]:
data2.source == 'SEO'

0        True
1        True
2       False
3       False
4       False
        ...  
1790    False
1791     True
1792    False
1793    False
1794    False
Name: source, Length: 1795, dtype: bool

Next, when we put this in the brackets, then we choose those rows of the `data2` DataFrame which had returned `True`. Thus, we get the output to be the following 346 rows:

# Chaining functions together:

As with normal Python, we can chain functions together to save memory and code, while compromising a little bit on readability.

Example:

Here we will first choose the first 5 rows and then choose some columns from that:

In [48]:
(data2.head(8))[['country','user_id','datetime']]

country     user_id             datetime
0  country_7  2458151261  2018-01-01 00:01:01
1  country_7  2458151262  2018-01-01 00:03:20
2  country_7  2458151263  2018-01-01 00:04:01
3  country_7  2458151264  2018-01-01 00:04:02
4  country_8  2458151265  2018-01-01 00:05:03
5  country_6  2458151266  2018-01-01 00:05:42
6  country_2  2458151267  2018-01-01 00:06:06
7  country_6  2458151268  2018-01-01 00:06:15

This same thing can also be done as follows: We first choose the columns we want, and then we take the first 5 rows from it. 

The output is exactly the same, another testimony to how everything in python can be done in more than one way

In [49]:
data2[['country','user_id','datetime']].head()

country     user_id             datetime
0  country_7  2458151261  2018-01-01 00:01:01
1  country_7  2458151262  2018-01-01 00:03:20
2  country_7  2458151263  2018-01-01 00:04:01
3  country_7  2458151264  2018-01-01 00:04:02
4  country_8  2458151265  2018-01-01 00:05:03

Now, one task for you all:

### Select the `user_id`, `topic` and `datetime` columns on the topic of North America, and print the last 15 rows only.

Hint for last part:

In [50]:
data2[['user_id','topic','datetime']][data2.topic=='North America'].tail(15)

user_id          topic             datetime
1689  2458152950  North America  2018-01-01 22:49:20
1694  2458152955  North America  2018-01-01 22:53:19
1706  2458152967  North America  2018-01-01 23:01:15
1716  2458152977  North America  2018-01-01 23:08:43
1721  2458152982  North America  2018-01-01 23:13:28
1723  2458152984  North America  2018-01-01 23:14:18
1741  2458153002  North America  2018-01-01 23:23:24
1743  2458153004  North America  2018-01-01 23:23:58
1751  2458153012  North America  2018-01-01 23:29:20
1754  2458153015  North America  2018-01-01 23:30:09
1756  2458153017  North America  2018-01-01 23:32:29
1762  2458153023  North America  2018-01-01 23:36:09
1777  2458153038  North America  2018-01-01 23:49:52
1784  2458153045  North America  2018-01-01 23:54:03
1790  2458153051  North America  2018-01-01 23:57:14

In [53]:
data2.loc[['user_id']]

KeyError: "None of [Index(['user_id'], dtype='object')] are in the [index]"